src/apps/chifra/pkg/abi/load_abi.go

```diff
@@ -0,0 +1,236 @@
+	}
+
+	for _, function := range functions {
+		function := function
+		destination[getMapKey(function.Encoding)] = &function
+	}
+	return true
+}
+
+func PreloadTokensAbi(destination AbiInterfaceMap) (err error) {
+	err = LoadAbiFromJsonFile(
+		path.Join(config.GetPathToRootConfig(), "abis", "known-000/erc_00020.json"),
+		destination,
+	)
+	if err != nil {
+		return
+	}
+	return LoadAbiFromJsonFile(
+		path.Join(config.GetPathToRootConfig(), "abis", "known-000/erc_00721.json"),
+		destination,
+	)
+}
+
+// PreloadKnownAbis loads known ABI files into destination, refreshing binary cache if needed
+func PreloadKnownAbis(chain string, destination AbiInterfaceMap, tokensOnly bool) (err error) {
+	if tokensOnly {
+		return PreloadTokensAbi(destination)
+	}
+
+	// Check if cache file is fresh
+	useCache, err := cache.IsAbiCacheUpToDate(chain)
+	if err != nil {
+		return
+	}
+	// Use cache file
+	if useCache {
+		if loaded := LoadCache(chain, destination); loaded {
+			return
+		}
+	}
+
+	// Use known .json files and freshen binary cache
+	paths, err := getKnownAbiPaths()
+	if err != nil {
+		return
+	}
+
+	for _, filePath := range paths {
+		loadErr := LoadAbiFromKnownFile(filePath, destination)
+		if loadErr != nil {
+			return fmt.Errorf("%s: %w", filePath, loadErr)
+		}
+	}
+
+	toCache := make([]types.SimpleFunction, 0, len(destination))
+	for _, function := range destination {
+		function := function
+		toCache = append(toCache, *function)
+	}
+	return cache.SetAbis(chain, toCache)
+}
+
+// findKnownAbi finds known ABI by name in known-* directories
+func findKnownAbi(name string) (filePath string) {
+	for _, subdirName := range knownAbiSubdirectories {
+		testPath := path.Join(config.GetPathToRootConfig(), "abis", subdirName, name+".json")
+		if file.FileExists(testPath) {
+			filePath = testPath
+			return
+		}
+	}
+	return
+}
+
+func getKnownAbiPaths() (filePaths []string, err error) {
+	knownDirPath := path.Join(config.GetPathToRootConfig(), "abis")
+	err = filepath.WalkDir(knownDirPath, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if d.IsDir() {
+			return nil
+		}
+		if filepath.Ext(d.Name()) != ".json" {
+			return nil
+		}
+		info, err := d.Info()
+		if err != nil {
+			return err
+		}
+		if info.Size() == 0 {
+			return nil
+		}
+
+		filePaths = append(filePaths, path)
+		return nil
+	})
+	return
+}
+
+// LoadAbiFromAddress loads ABI from local file or cache
+func LoadAbiFromAddress(chain string, address common.Address, destination AbiInterfaceMap) (err error) {
+	localFileName := strings.ToLower(address.Hex()) + ".json"
+	localFile, err := os.OpenFile(localFileName, os.O_RDONLY, 0)
+	if os.IsNotExist(err) {
+		// There's no local file, so we try to load one from cache
+		loadedAbis, err := cache.GetAbi(chain, address)
+		if err != nil {
+			return err
+		}
+		for _, loadedAbi := range loadedAbis {
+			loadedAbi := loadedAbi
+			destination[getMapKey(loadedAbi.Encoding)] = &loadedAbi
```

Comment:
Is it possible to make sure the Encoding is already lower case when it's loaded? Not sure that makes sense, just curious.